src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,199 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+            // sending range requests to the Razor/C#/HTML servers.
+            var filledInRange = await FillInRangeGapsAsync(
+                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                range, cachedRange, cachedTokens, cancellationToken).ConfigureAwait(false);
+
+            return filledInRange;
+        }
+
+        private async Task<SemanticTokens?> FillInRangeGapsAsync(
+            TextDocumentIdentifier textDocumentIdentifier,
+            DocumentSnapshot documentSnapshot,
+            int documentVersion,
+            VersionStamp semanticVersion,
+            Range range,
+            Range cachedRange,
+            List<int> cachedTokens,
+            CancellationToken cancellationToken)
+        {
+            var tokenList = new List<int>();
+            var absoluteLine = 0;
+
+            // There are 3 scenarios:
+            //    1) The cached result is at the start of the range
+            //    2) The cached result is in the middle of the range
+            //    3) The cached result is at the end of the range
+            // For 1) and 3), we need to send one request to the Razor/C# language servers for the missing range's tokens.
+            // For 2), we need to send two requests (one for the missing start range, and one for the missing end range).
+
+            // Before processing the cached portion of the range, we might have to compute tokens for the
+            // start of the range -> start of the cached range.
+            if (cachedRange.Start.Line != range.Start.Line)
+            {
+                var partialRange = new Range { Start = range.Start, End = cachedRange.Start };
+                var tokens = await ComputeAndCacheTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                    partialRange, cancellationToken).ConfigureAwait(false);
+
+                // Add the partially computed range to the results. We need to also keep track of the current
+                // absolute line index to use later when combining results.
+                var firstToken = true;
+                for (var tokenIndex = 0; tokenIndex < tokens.Data.Length; tokenIndex += 5)
+                {
+                    absoluteLine += tokens.Data[tokenIndex];
+
+                    // Skip tokens that are out of the range
+                    if (absoluteLine < range.Start.Line || absoluteLine >= cachedRange.Start.Line)
+                    {
+                        continue;
+                    }
+
+                    if (firstToken)
+                    {
+                        tokenList.Add(absoluteLine);
+                        firstToken = false;
+                    }
+                    else
+                    {
+                        tokenList.Add(tokens.Data[tokenIndex]);
+                    }
```

Comment:
fwiw, i can't reason about this code at all.  it's just super confusing.  esp with the embedded 'continue', and then the boolean check of firstToken.  

First, i think it would make a ton more sense to just have a single routine compute the very first value (since htat seems to be the special line-offset value, then the rest is just a simple copy.

Second, this feature area suffers from what i think of as working with the wrong abstraction.  Yes, semantics tokens are a 5-tuple of ints with meaning.  But the continual work to process everything a 5x array of ints where the code must be incredibly delicate with that, is really unpleasant.  

This area would really benefit from taking that int-array and wrapping it behind some other structure that exposes a much more pleasant and easy to read API for things like the service to use.  For example .MoveLineToNewStartPosition (or something like that).

Internally, we can then have very low level operations (like this) which doc the structure of the int[] and what we're doing with it.

Reply:
> First, i think it would make a ton more sense to just have a single routine compute the very first value (since htat seems to be the special line-offset value, then the rest is just a simple copy.

Totally get what you mean. However, in this specific case, I don't think it's possible to break out the computation of the first value since it's not technically always the first value, but rather the "first valid value in the range". I added a comment + renamed the variable to reflect this. There are a couple of other places in the code where I did make changes to reflect the pattern you suggested however.

> This area would really benefit from taking that int-array and wrapping it behind some other structure that exposes a much more pleasant and easy to read API for things like the service to use. For example .MoveLineToNewStartPosition (or something like that).

In my next push I've added a ton of helper methods to hopefully make readability easier. For now, I personally don't think the added methods make as much sense in an API (besides the methods already in SemanticTokensCache), but if you have a different opinion lmk!